### PR TITLE
ES|QL: Fix DISSECT that overwrites input

### LINKED ISSUE #110184: ES|QL: VerificationException with DISSECT masking input
A DISSECT statement can produce an attribute with the same name as the input, eg.

```
FROM idx | DISSECT foo "%{foo} %{bar}" | KEEP foo, bar
```

If the query is followed by a KEEP, the result is a VerificationException: "Unknown column [foo]".
Without the KEEP the query works fine.

The problem seems to be in [EsqlSession.fieldNames()](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java#L243-L245): when the input name clashes with the output names, the input is also wrongly removed from the list and hence not requested for index resolution. 
Without the KEEP, the index resolution falls back to ALL_FIELDS, so the problem does not happen.

### PR DESCRIPTION
**Pull request by luigidellaquila**
Fixes https://github.com/elastic/elasticsearch/issues/110184

When a DISSECT command overwrites the input, eg.

```
FROM idx | DISSECT foo "%{foo} %{bar}" | KEEP foo, bar
```

The input field (`foo` in this case) could be excluded from the index resolution (incorrectly masked by the `foo` that is the result of the DISSECT).
This PR makes sure that the input field does not get lost and is correctly passed to the indexResolver.

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/dissect.csv-spec [hunk 1]
```diff
@@ -175,6 +175,14 @@ Parto Bamford     | Parto          | Bamford
 ;
 
 
+overwriteInputName
+from employees | sort emp_no asc | dissect first_name "%{first_name}o%{rest}" | keep emp_no, first_name, rest | limit 1;
```
Comment (alex-spies) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/dissect.csv-spec:179]: Weird. There's a test already for this case, `shadowingSelf`. Maybe we should add a unit test for the analyzer to test explicitly what's going on?

At least, let's add a comment that says why this is not redundant with `shadowingSelf`.

**Review by alex-spies — APPROVED** [review]
LGTM, thanks Luigi!

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/dissect.csv-spec (hunk 1 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/dissect.csv-spec:179]: The difference is that `shadowingSelf` dissects the result of an EVAL, not a field that is loaded straight from an index, so the analyzer is not impacted.
I'll add a comment to clarify the scope of the second test (and I'll link the issue) 👍

**Review by astefan — APPROVED** [review]
LGTM
I would also add a test that shows that the type of the field that has the same name is the correct one that comes out of the RegexExtract:
`row text = "123 abc", int = 5 | sort int asc | grok text "%{NUMBER:text:int} %{WORD:description}"`

**Comment by luigidellaquila** [comment]
Thanks for the feedback folks!

**Pull request merged**